### ISSUE TITLE
As a performance engineer I want RFSwarm to add think times and pacing by configuration

### ISSUE THREAD

**Issue by dschweie** [damies13/rfswarm#174]
The unique selling point of RFSwarm is the reuse of existing functional test cases for performance testing. In order to be able to take advantage of this advantage more clearly, the necessary breaks should not be incorporated manually into the test case. Through configuration it should be possible to make specifications, which are then incorporated into the test case by RFSwarm.

This requirement must definitely be seen and evaluated in conjunction with #172.

**Comment by damies13** [comment]
Hi Dirk,

Thanks for the enhancement request, I remember our conversation about this, but I may not in future, so these are my notes to remind myself as much as anything, correct me if I misunderstood.

_**anyone else** Please add comments if you wish_

My initial thought on this is to add an option to the scenario settings and test row settings to add sleep keywords:
- checkbox to enable
- time in seconds (maybe we can eliminate the checkbox by having 0 to disable?)
- My initial thought is this should only apply to measured keywords (those reported on the run screen)
  - Should there be an option to apply to all keywords (even the ones not measured)?
  - Should there be an option to control keyword levels? e.g. if you are measuring the time to fill a form but also the time for the individual tabs or sub pages would you want the sleep added to the tab loads too?
  - Should we take advantage of RFv7.0 features and have an option to apply the sleep to only user keywords, only library keywords, and both? This might mean the feature will only be available to agents with RF >= v7.0, I've never placed this type of restriction before, always tried to have RFSwarm work with all supported RF versions.

The way I expect to be able to implement this on the agent is to add a trigger to do a sleep in the end_keyword of the existing listener, will have to see if that will work.

**Comment by damies13** [comment]
Had a bit of a play with the V3 listener today, It's not ideal, but I think I have a workable solution.

```
				tmeslp = 2
				time.sleep(tmeslp)
				data.parent.body.create_keyword("Log", ["RFSwarm Agent triggered a sleep of {} seconds afer {}".format(tmeslp, data.name)])
```

I can add the `time.sleep(tmeslp)` to the end of end_keyword in both v2 and v3 listener, but the robot framework log.html has unexplained missing time, each keyword will start n seconds after the previous one finished with no impact on the elapsed time of the keywords but there is an unexplained increase to the elapsed time of the parent keyword/test/for, etc

With the v2 Listener there is nothing I could do about it for now
With the v3 Listener I experimented with data.parent.body.create_keyword, initially creating a sleep keyword, but rather than adding the sleep after the keyword that just executed it added the keyword to the end of the parent keyword/test/for, etc, so the solution I'm thinking to go with is to append a log with a note that RFSwarm Agent added a sleep with the time slept and the keyword that triggered it.

Unfortunately I'll only be able to have this not to explain the missing time on the v3 Listener, so that means RF7+, RF6.x and lower will just have to have missing time in the logs, not sure if that's going to cause issues or not.

**Comment by damies13** [comment]
Example of the V3 log
![image](https://github.com/damies13/rfswarm/assets/15232871/21b7fe97-221f-4391-9463-3cbca51e8138)

**Comment by dschweie** [comment]
In my opinion, this is going in the right direction. It would be very helpful if I, as a performance engineer, could configure the pause times and possible pacing times per keyword.

**Comment by damies13** [comment]
Yes, the intention is to add configuration to the settings screens in the manager, this part was making sure I could get it working in the listener first as adding items in the settings screens is the easy  part `tmeslp = 2` we just to represent the value received from the manager.

**Comment by damies13** [comment]
Started implementing this in the manager GUI, here's a little preview of the test defaults setting screen:

![image](https://github.com/damies13/rfswarm/assets/15232871/03def0a6-5ef3-4b23-9dff-090bde3aa650)

The plan is have a minimum and maximum value for sleep and agent listener will pick a random value between those, but if you want the sleeps to always be the same just set min and max to the same number.

I set the default min to 15 and default max to 45 seconds giving a default average of 30 seconds.

I wasn't sure what to call this, the names I tossed around were:
- Insert Sleep
- Insert Pacing
- Insert Think Time
- Inject Sleep
- Inject Pacing
- Inject Think Time

In the end I went with "Inject Sleep" as Sleep is the keyword name, and I think inject might be the more appropriate description, as in: injecting code, which probably more accurately describes what this is doing.

https://wikidiff.com/inject/insert

It's not too late to change it, so happy to discuss
